p2p/sharding.py

```diff
@@ -0,0 +1,209 @@
+import asyncio
+import logging
+import random
+import time
+from typing import (
+    List,
+)
+
+import rlp
+
+from evm.rlp.collations import Collation
+from evm.rlp.headers import CollationHeader
+from evm.chains.shard import Shard
+
+from evm.db.shard import (
+    Availability,
+)
+
+from evm.constants import (
+    COLLATION_SIZE,
+)
+from evm.exceptions import (
+    CanonicalCollationNotFound,
+)
+
+from p2p.cancel_token import (
+    CancelToken,
+    wait_with_token,
+)
+from p2p import protocol
+from p2p.protocol import (
+    Command,
+    Protocol,
+)
+from p2p.peer import (
+    BasePeer,
+    PeerPool,
+    PeerPoolSubscriber,
+)
+from p2p.p2p_proto import (
+    DisconnectReason,
+)
+from p2p.exceptions import (
+    HandshakeFailure,
+)
+
+
+COLLATION_PERIOD = 1
+
+
+class Status(Command):
+    _cmd_id = 0
+
+
+class Collations(Command):
+    _cmd_id = 1
+
+    structure = rlp.sedes.CountableList(Collation)
+
+
+class ShardingProtocol(Protocol):
+    name = "sha"
+    version = 0
+    _commands = [Status, Collations]
+    cmd_length = 2
+
+    logger = logging.getLogger("p2p.sharding.ShardingProtocol")
+
+    def send_handshake(self) -> None:
+        cmd = Status(self.cmd_id_offset)
+        self.logger.debug("Sending status msg")
+        self.send(*cmd.encode([]))
+
+    def send_collations(self, collations: List[Collation]) -> None:
+        cmd = Collations(self.cmd_id_offset)
+        self.logger.debug("Sending {} collations".format(len(collations)))
+        self.send(*cmd.encode(collations))
+
+
+class ShardingPeer(BasePeer):
+    _supported_sub_protocols = [ShardingProtocol]
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.incoming_collation_queue = None
+        self.known_collation_hashes = set()
+
+    def set_incoming_collation_queue(self, queue: asyncio.Queue) -> None:
+        self.incoming_collation_queue = queue
+
+    #
+    # Handshake
+    #
+    async def send_sub_proto_handshake(self) -> None:
+        self.sub_proto.send_handshake()
+
+    async def process_sub_proto_handshake(
+        self,
+        cmd: Command,
+        msg: protocol._DecodedMsgType
+    ) -> None:
+        if not isinstance(cmd, Status):
+            self.disconnect(DisconnectReason.other)
+            raise HandshakeFailure("Expected status msg, got {}, disconnecting".format(cmd))
+
+    #
+    # Receiving Collations
+    #
+    def handle_sub_proto_msg(self, cmd: Command, msg: protocol._DecodedMsgType) -> None:
+        if isinstance(cmd, Collations):
+            self._handle_collations_msg(msg)
+        else:
+            super().handle_sub_proto_msg(cmd, msg)
+
+    def _handle_collations_msg(self, msg: List[Collation]) -> None:
+        for collation in msg:
+            self.known_collation_hashes.add(collation.hash)
+            if self.incoming_collation_queue is not None:
```

Comment:
This warrants a comment explanation.  Why is this *sometimes* `None` and how does it get set.